### ISSUE TITLE
WalletTest#basicSpendingWithEncryptedWallet issue

### ISSUE THREAD

**Issue by zkejid** [bitcoinj/bitcoinj#1410]
Hello.

Today I found a bug during Travis build. If I configure local test environment run WalletTest#basicSpendingWithEncryptedWallet in endless loop until failure, once it stops due to exception. Fail case is pretty rare: usually there are about one thousand successful calls before failure, but sometimes it happens nearly after start. 

Code base:

> 89fb8c3 Andreas Schildbach <andreas.schildbach@bloq.com> on 19.07.17 at 0:21

Environment:

> java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) Server VM (build 25.131-b11, mixed mode)

> Ubuntu 16.04.2 LTS

Can you suggest the origin cause of exception?

Here is the log:
```
19:01:21 1 Context.<init>: Creating bitcoinj 0.15-SNAPSHOT context.
19:01:21 1 MnemonicCode.toSeed: PBKDF2 took 89.93 ms
19:01:21 1 KeyChainGroup.addAndActivateHDChain: Creating and activating a new HD chain: org.bitcoinj.wallet.DeterministicKeyChain@c4944a
19:01:21 1 DeterministicKeyChain.maybeLookAhead: 1 keys needed for M/0H/0 = 1 issued + 0 lookahead size + 0 lookahead threshold - 0 num children
19:01:21 1 DeterministicKeyChain.maybeLookAhead: Took 193.7 μs
19:01:21 1 AbstractBlockChain.<init>: chain head is at height 0:
 block: 
   hash: 6cfd0819a94be376f70b36d3df1476e15b64c2aa9aed4ffe75b6f93381a2676a
   version: 1
   previous block: 0000000000000000000000000000000000000000000000000000000000000000
   merkle root: 4a5e1e4baab89f3a32518a88c31bc87f618f76673e2cc77ab2127b7afdeda33b
   time: 1500724711 (2017-07-22T11:58:31Z)
   difficulty target (nBits): 545259519
   nonce: 0

19:01:21 1 MnemonicCode.toSeed: PBKDF2 took 91.62 ms
19:01:21 1 KeyChainGroup.addAndActivateHDChain: Creating and activating a new HD chain: org.bitcoinj.wallet.DeterministicKeyChain@e39da0
19:01:21 1 KeyCrypterScrypt.deriveKey: Deriving key took 194.7 ms for 16384 scrypt iterations.
19:01:21 1 DeterministicKeyChain.maybeLookAhead: 1 keys needed for M/0H/0 = 1 issued + 0 lookahead size + 0 lookahead threshold - 0 num children
19:01:21 1 DeterministicKeyChain.maybeLookAhead: Took 195.8 μs
19:01:21 1 Wallet.receivePending: Received a pending transaction 82b7416d82894ae0a440da3523118da96b02921c78af60f90e8c9f7ffdf1ccdb that spends 0.00 BTC from our own wallet, and sends us 1.00 BTC
19:01:21 1 Wallet.receivePending: Wallet received transaction with an unknown source. Consider tagging it!
19:01:21 1 Wallet.maybeCommitTx: commitTx of 82b7416d82894ae0a440da3523118da96b02921c78af60f90e8c9f7ffdf1ccdb
19:01:21 1 Wallet.maybeCommitTx: ->pending: 82b7416d82894ae0a440da3523118da96b02921c78af60f90e8c9f7ffdf1ccdb
19:01:21 1 Wallet.maybeCommitTx: Estimated balance is now: 1.00 BTC
19:01:21 1 Wallet.receive: Received tx for 1.00 BTC: 82b7416d82894ae0a440da3523118da96b02921c78af60f90e8c9f7ffdf1ccdb [0] in block 4536c2e99003f70082dc0491592357fed90adc69609b1c0282707203e06d4c62
19:01:21 1 Wallet.receive:   <-pending
19:01:21 1 Wallet.processTxFromBestChain:   tx 82b7416d82894ae0a440da3523118da96b02921c78af60f90e8c9f7ffdf1ccdb ->unspent
19:01:21 1 Wallet.receive: Balance is now: 1.00 BTC
19:01:21 1 Wallet.completeTx: Completing send tx with 1 outputs totalling 10.00 BTC and a fee of 0.00 BTC/kB
19:01:22 1 KeyCrypterScrypt.deriveKey: Deriving key took 196.5 ms for 16384 scrypt iterations.
19:01:22 1 KeyCrypterScrypt.deriveKey: Deriving key took 206.0 ms for 16384 scrypt iterations.
19:01:22 1 Wallet.completeTx: Completing send tx with 1 outputs totalling 0.50 BTC and a fee of 0.00 BTC/kB
19:01:22 1 DeterministicKeyChain.maybeLookAhead: 1 keys needed for M/0H/1 = 1 issued + 0 lookahead size + 0 lookahead threshold - 0 num children
19:01:22 1 DeterministicKeyChain.maybeLookAhead: Took 215.9 μs
19:01:22 1 Wallet.completeTx:   with 0.50 BTC change
19:01:22 1 Wallet.completeTx: Completing send tx with 1 outputs totalling 0.50 BTC and a fee of 0.00 BTC/kB
19:01:22 1 Wallet.completeTx:   with 0.50 BTC change

java.lang.IllegalArgumentException: private key exceeds 32 bytes: {} bits [376]

	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:180)
	at org.bitcoinj.core.ECKey.<init>(ECKey.java:193)
	at org.bitcoinj.crypto.DeterministicKey.<init>(DeterministicKey.java:69)
	at org.bitcoinj.crypto.DeterministicKey.derivePrivateKeyDownwards(DeterministicKey.java:427)
	at org.bitcoinj.crypto.DeterministicKey.findOrDeriveEncryptedPrivateKey(DeterministicKey.java:406)
	at org.bitcoinj.crypto.DeterministicKey.decrypt(DeterministicKey.java:377)
	at org.bitcoinj.crypto.DeterministicKey.decrypt(DeterministicKey.java:43)
	at org.bitcoinj.core.ECKey.decrypt(ECKey.java:1113)
	at org.bitcoinj.crypto.DeterministicKey.decrypt(DeterministicKey.java:388)
	at org.bitcoinj.crypto.DeterministicKey.decrypt(DeterministicKey.java:43)
	at org.bitcoinj.wallet.DecryptingKeyBag.maybeDecrypt(DecryptingKeyBag.java:50)
	at org.bitcoinj.wallet.DecryptingKeyBag.findKeyFromPubHash(DecryptingKeyBag.java:67)
	at org.bitcoinj.core.TransactionOutPoint.getConnectedRedeemData(TransactionOutPoint.java:169)
	at org.bitcoinj.core.TransactionInput.getConnectedRedeemData(TransactionInput.java:295)
	at org.bitcoinj.wallet.Wallet.signTransaction(Wallet.java:4091)
	at org.bitcoinj.wallet.Wallet.completeTx(Wallet.java:4028)
	at org.bitcoinj.wallet.WalletTest.basicSpendingCommon(WalletTest.java:340)
	at org.bitcoinj.wallet.WalletTest.basicSpendingWithEncryptedWallet(WalletTest.java:164)
	at sun.reflect.GeneratedMethodAccessor2.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:67)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

**Comment by schildbach** [comment]
Thanks for catching this. Most likely this was introduced in my commit 3890219540048b21d59ef721f84725a30b50e84f (PR #1373). As a first step, I just fixed the exception message so that it contains the actual number of bits the private key has.

But I suspect we will undo my commit entirely. @natzei's commit 3cb1f6c6c589f84fe6e1fb56bf26d94cccc85429 (PR #1404) perhaps makes it obsolete.

**Comment by schildbach** [comment]
I just reproduced this, the message is "private key exceeds 32 bytes: 375 bits". In your (above) case it seems to have been 376 bits.

**Comment by schildbach** [comment]
Just tested on 0.14.4, cherry-picking 3890219 (the failing constraint) and the issue exists there, too.

**Comment by schildbach** [comment]
Ok, after some hours of investigation I think we uncovered a subtle but serious bug with wallet encryption here. It only happens when you supply the *wrong* password. Normally, this throws up a KeyCrypterException("Could not decrypt bytes"). But in this case, it decrypts to a (wrong) deterministic key m/0'/0 with a length of around 47 bytes (rather than 32). This triggers the new sanity check (PR #1373), for good.

Now, I don't want to falsely panic, but I think this can have severe consequences for encrypted wallets if the users enters the wrong password and wallet encryption doesn't detect this properly. Because further private keys are derived from the m/0'/0 key. I'm not sure if any if these private keys can get saved to the wallet (normally they're not because they get re-derived on demand). I'm also not sure if this can affect public keys too. If yes, it could lead to receiving coins on wrong addresses.

The good news is, the new sanity check seems to contain the damage (not for 0.14.4 though).

**Comment by systemsplanet** [comment]
FYI.
I've just had a unit test fail from this error:
```
[INFO] encryptionDecryptionBadPassword(org.bitcoinj.wallet.WalletTest)  Time elapsed: 0.164 sec  <<< ERROR!
[INFO] java.lang.IllegalArgumentException: private key exceeds 32 bytes: 376 bits
[INFO] 	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:181)
[INFO] 	at org.bitcoinj.core.ECKey.<init>(ECKey.java:183)
[INFO] 	at org.bitcoinj.core.ECKey.<init>(ECKey.java:178)
[INFO] 	at org.bitcoinj.crypto.DeterministicKey.<init>(DeterministicKey.java:78)
[INFO] 	at org.bitcoinj.crypto.DeterministicKey.decrypt(DeterministicKey.java:337)
[INFO] 	at org.bitcoinj.crypto.DeterministicKey.decrypt(DeterministicKey.java:44)
[INFO] 	at org.bitcoinj.core.ECKey.decrypt(ECKey.java:1083)
[INFO] 	at org.bitcoinj.crypto.DeterministicKey.decrypt(DeterministicKey.java:345)
[INFO] 	at org.bitcoinj.wallet.DeterministicKeyChain.checkAESKey(DeterministicKeyChain.java:1061)

[INFO] 	at org.bitcoinj.wallet.KeyChainGroup.checkAESKey(KeyChainGroup.java:336)
[INFO] 	at org.bitcoinj.wallet.Wallet.checkAESKey(Wallet.java:1260)
[INFO] 	at org.bitcoinj.wallet.WalletTest.encryptionDecryptionBadPassword(WalletTest.java:1864)
[INFO] 
[INFO] 
[INFO] Results :
[INFO] 
[INFO] Tests in error: 
[INFO]   WalletTest.encryptionDecryptionBadPassword:1864 ? IllegalArgument private key ...
```

**Comment by schildbach** [comment]
The test error in `WalletTest.encryptionDecryptionBadPassword()` should be fixed via  c5348d2c216f1644c169b042ddda048837c194fb by checking the decrypted key size and throwing a `KeyCrypterException.InvalidCipherText` exception in this case. Which is correctly handled by the test case (is expected to happen).

**Comment by schildbach** [comment]
Fixed the original issue too, via 06d9b6489060584fa4c9b455b0a5856f3348953f.
